curated_transformers/util.py

```diff
@@ -43,3 +48,81 @@ def all_equal(iterable):
     (or if the input is an empty sequence), False otherwise."""
     g = itertools.groupby(iterable)
     return next(g, True) and not next(g, False)
+
+
+def gradual_transformer_unfreezing_per_pipe(
+    nlp: Language,
+    callback_args: Dict[str, Any],
+    freeze_params: Dict[str, int],
+    last_unfreeze_step: int,
+):
+    current_step = callback_args["step"]
+    if current_step > last_unfreeze_step:
+        return
+
+    # Scoped import to avoid import cycles.
+    from .pipe import Transformer
+
+    for name, pipe in nlp.components:
+        unfreeze_step = freeze_params.get(name)
+        if unfreeze_step is None:
+            continue
+        elif not isinstance(pipe, Transformer):
+            raise TypeError(
+                "Gradual unfreezing cannot be performed on non-Transformer component '{name}'"
+            )
+
+        _handle_transformer_freezing(pipe, current_step, unfreeze_step)
+
+
+def gradual_transformer_unfreezing_all_pipes(
+    nlp: Language, callback_args: Dict[str, Any], unfreeze_step: int
+):
+    current_step = callback_args["step"]
+    if current_step > unfreeze_step:
+        return
+
+    # Scoped import to avoid import cycles.
+    from .pipe import Transformer
+
+    for _, pipe in nlp.components:
+        if not isinstance(pipe, Transformer):
+            continue
+
+        _handle_transformer_freezing(pipe, current_step, unfreeze_step)
+
+
+def _handle_transformer_freezing(
+    pipe: "Transformer", current_step: int, unfreeze_step: int
+):
+    if current_step == 0:
+        # Always freeze at the start of training.
+        pipe.frozen = True
+    elif current_step < unfreeze_step:
+        assert (
+            pipe.frozen == True
+        ), "Transformer frozen state unexpectedly changed before unfreezing"
+    elif current_step == unfreeze_step:
+        pipe.frozen = False
```

Comment:
```suggestion
    pipe.frozen = current_step < unfreeze_step
```

I don't think it's good to have the assertion here, maybe the user made a composed callback?